private/buf/bufgen/generator.go

```diff
@@ -76,49 +89,215 @@ func (g *generator) generate(
 	image bufimage.Image,
 	baseOutDirPath string,
 	includeImports bool,
+	printFormat bufprint.Format,
 ) error {
 	if err := modifyImage(ctx, config, image); err != nil {
 		return err
 	}
-	// We keep this as a variable so we can cache it if we hit StrategyDirectory.
+	// Cache imagesByDir up-front if at least one plugin requires it
 	var imagesByDir []bufimage.Image
-	var err error
 	for _, pluginConfig := range config.PluginConfigs {
-		out := pluginConfig.Out
-		if baseOutDirPath != "" && baseOutDirPath != "." {
-			out = filepath.Join(baseOutDirPath, out)
+		if pluginConfig.Strategy != StrategyDirectory {
+			continue
 		}
-		var pluginImages []bufimage.Image
-		switch pluginConfig.Strategy {
-		case StrategyAll:
-			pluginImages = []bufimage.Image{image}
-		case StrategyDirectory:
-			// If we have not already called this, call it.
-			if imagesByDir == nil {
-				imagesByDir, err = bufimage.ImageByDir(image)
-				if err != nil {
-					return err
-				}
-			}
-			pluginImages = imagesByDir
-		default:
-			return fmt.Errorf("unknown strategy: %v", pluginConfig.Strategy)
+		// Already guaranteed by config validation, but best to be safe.
+		if pluginConfig.Remote != "" {
+			return fmt.Errorf("remote plugin %q cannot set strategy directory", pluginConfig.Remote)
+		}
+		var err error
+		imagesByDir, err = bufimage.ImageByDir(image)
+		if err != nil {
+			return err
 		}
-		if err := g.appprotoosGenerator.Generate(
-			ctx,
-			container,
-			pluginConfig.Name,
-			out,
-			bufimage.ImagesToCodeGeneratorRequests(
+		break
+	}
+	remoteToPlugins := make(map[string][]remotePluginOptions)
+	var localPluginResponses []localPluginResponse
+	// Parallelize local and remote plugin execution with an errgroup.
+	// Note: not using thread.Parallelize because we want to be able to
+	// cancel all execution if one plugin errors. Limit concurrent execution
+	// of local plugins using semaphore channel.
+	semaphoreC := make(chan struct{}, thread.Parallelism())
+	defer close(semaphoreC)
+	eg, ctx := errgroup.WithContext(ctx)
```

Comment:
We use `thread.Parallelize`. As discussed, if we need to extend this to have a `ctx`, we can

Reply:
Discussed offline, after improvements in #502 I will update this to use `thread.Parallelize`.